core/trino-main/src/main/java/io/trino/sql/ExpressionUtils.java

```diff
@@ -265,6 +276,40 @@ public static Expression filterConjuncts(Metadata metadata, Expression expressio
         };
     }
 
+    /**
+     * Returns whether expression is effectively literal. An effectitvely literal expression is a constant value, or null,
+     * in either {@link Literal} form, or other form returned by {@link LiteralEncoder}.
+     */
+    public static boolean isEffectivelyLiteral(PlannerContext plannerContext, Session session, Expression expression)
+    {
+        if (expression instanceof Literal) {
+            return true;
+        }
+        if (expression instanceof Cast) {
+            return ((Cast) expression).getExpression() instanceof Literal
+                    // a Cast(Literal(...)) can fail, so this requires verification
+                    && simplifiesToConstant(plannerContext, session, expression);
+        }
+        if (expression instanceof FunctionCall) {
+            QualifiedName functionName = ((FunctionCall) expression).getName();
+            if (isResolved(functionName)) {
+                ResolvedFunction resolvedFunction = plannerContext.getMetadata().decodeFunction(functionName);
+                return LITERAL_FUNCTION_NAME.equals(resolvedFunction.getSignature().getName());
+            }
+        }
+
+        return false;
+    }
+
+    private static boolean simplifiesToConstant(PlannerContext plannerContext, Session session, Expression expression)
```

Comment:
`ExpressionInterpreter.evaluateConstantExpression()` that could be useful here.

Reply:
No, just consider. I haven't dug into how appropriate it is for this purpose.

Reply:
`evaluateConstantExpression` is not appropriate here, because it throws if it cannot evaluate (eg a failing cast).
i don't want the logic here to be exception-driven

OTOH, i like the idea `evaluateConstantExpression`  so i used it in one place (where i already know something is a "effectively literal"). In the future, for wider use of the method, we should have a version of `evaluateConstantExpression` that isn't scarily long, with all these coercions, and 5x re-analysis of an expression.